neuralpp/symbolic/sympy_interpreter.py

```diff
@@ -35,17 +36,62 @@ def context_to_variable_value_dict(context: Expression) -> \
             return result
         else:
             raise RuntimeError(f"cannot evaluate to a value. The best effort result is {result}.")
 
-    def simplify(self, expression: SymPyExpression, context: SymPyExpression) -> SymPyExpression:
-        raise NotImplementedError("TODO")
+    @staticmethod
+    def _updated_type_dict_by_context(type_dict: Dict[sympy.Basic, ExpressionType], context: Expression) -> \
+            Dict[sympy.Basic, ExpressionType]:
+        """ argument type_dict is not modified, instead, a new dict is created as a return value"""
+        result = {}
+        for k, v in type_dict.items():
+            new_key = SymPyInterpreter._simplify_expression(k, context)
+            if new_key == k or new_key not in type_dict:
+                # if skip the update if not(new_key == k or new_key not in type_dict), i.e,
+                # if (i) new_key is a new key (has been simplified()) and (ii) new_key exists in type_dict.
+                # From (i) and (ii), k must be a function application (if k is a variable and simpified, new_key would
+                # be constant and will not be in type_dict). Based on the type of new_key, there's two cases:
+                # #1 if new_key is a variable: then new_key should be discarded as v is of function type.
+                # #2 if new_key is a function application: old and new types should be same. just discard the new_key.
+                result[new_key] = v
+        return result
+
+    def simplify(self, expression: Expression, context: Expression = BasicConstant(True)) -> SymPyExpression:
+        """
+        The function calls simplify() from sympy library and wrap the result in SymPyExpression.
+        """
+        if not isinstance(expression, SymPyExpression):
+            expression = SymPyExpression.convert(expression)
+
+        result = SymPyInterpreter._simplify_expression(expression.sympy_object, context)
+        # if "x+y" is simplified to "x+2", how do we find out the type of that "+"? We need to update the type_dict
+        # by also doing the replacement. If we have "x+(y+y)", the type_dict would be
+        #   {x:int, y:int, y+y:int->int->int, x+(y+y):int->int->int}
+        # by a context of {y:2}, we'd have
+        #   {x:int, 2:int, 4:int->int->int, x+4:int->int->int} (entry 2 and 4 will be GCed)
+        type_dict = SymPyInterpreter._updated_type_dict_by_context(expression.type_dict, context)
+        type_ = type_dict[result]
+        result_expression = SymPyExpression.from_sympy_object(result, type_, type_dict)
+        # The result keeps the known type information from `expression`. E.g., though (y-y).simplify() = 0, it still
+        # keeps the type of `y`. Say the old `y` is of int type, adding that `0` to another `y` of float type will
+        # cause an error. That's why we need garbage_collect_type_dict() for cleaning the old, unreachable type
+        # information.
+        result_expression.garbage_collect_type_dict()
```

Comment:
Sorry for being late to the party. I have a dumb question regarding `garbage_collect_type_dict` and how it's being used here: IIUC the `Expression` class is designed to be immutable, and in most other methods, we always return a new instance of the expression instead of doing in-place modification on `self`.

I understand the motivation to "prune" unused type dict entries, though I'm wondering if pruning the dict on `self` is slightly violating the immutable design? Happy to hear about your thoughts

Reply:
My argument is that

1. `result_expression` created at line #91 is still "temporary" and not visible to the caller, so it's virtually immutable; and
2. immutability in programming languages is more or less an virtualization/illusion (e.g., C++ `const` method can still change the state of the object), especially in python which provides no way to enforce that.

That being said, for this specific example, I can change the design a bit to adhere better to the immutability principle. We can "prune" the `type_dict` first and then return `from_sympy_object()`, essentially swapping line #91 and #96.